src/main/anovos/data_transformer/transformers.py

```diff
@@ -3430,39 +3431,46 @@ def outlier_categories(
         for index, i in enumerate(list_of_cols):
             window = Window.partitionBy().orderBy(F.desc("count_pct"))
             df_cats = (
-                idf.groupBy(i)
-                .count()
-                .dropna()
-                .withColumn(
-                    "count_pct",
-                    F.col("count") / F.sum("count").over(Window.partitionBy()),
-                )
-                .withColumn("rank", F.rank().over(window))
-                .withColumn(
-                    "cumu",
-                    F.sum("count_pct").over(
-                        window.rowsBetween(Window.unboundedPreceding, 0)
-                    ),
-                )
-                .withColumn("lag_cumu", F.lag("cumu").over(window))
-                .fillna(0)
-                .where(~((F.col("cumu") >= coverage) & (F.col("lag_cumu") >= coverage)))
-                .where(F.col("rank") <= (max_category - 1))
-                .select(F.lit(i).alias("attribute"), F.col(i).alias("parameters"))
+                idf.select(i).groupBy(i)
+                    .count()
+                    .dropna()
+                    .withColumn(
+                        "count_pct",
+                        F.col("count") / F.sum("count").over(Window.partitionBy()),
+                    )
+                    .withColumn("rank", F.rank().over(window))
+                    .withColumn(
+                        "cumu",
+                        F.sum("count_pct").over(
+                            window.rowsBetween(Window.unboundedPreceding, 0)
+                        ),
+                    )
+                    .withColumn("lag_cumu", F.lag("cumu").over(window))
+                    .fillna(0)
+                    .where(~((F.col("cumu") >= coverage) & (F.col("lag_cumu") >= coverage)))
+                    .where(F.col("rank") <= (max_category - 1))
+                    .select(F.lit(i).alias("attribute"), F.col(i).alias("parameters"))
             )
             if index == 0:
                 df_model = df_cats
             else:
                 df_model = df_model.union(df_cats)
 
+    df_params = df_model.rdd.groupByKey().mapValues(list).collect()
+    dict_params = dict(df_params)
+    broadcast_params = spark.sparkContext.broadcast(dict_params)
```

Comment:
Have you tested the runtime with and without `broadcast`? Also the utilization of memory and resource as well. Since I think we only use 1 machine here, I'm not sure how `broadcast` can help.

Reply:
I am not sure why you are saying that we only use 1 machine here ? 
broadcast variables are shared across all executers. 
Why Spark Broadcast is used?
When huge datasets are needed to be cached in executors, the broadcast variables are used. Let us imagine, we need to lookup for pin codes or zip codes for doing a transformation and in this case, it is neither we will be able to query the huge database every time nor it is practical to send the huge lookup for the table every time to the executors. The only solution is by converting the table lookup into a broadcast variable and will be cached by Spark in every executor in reference for future use.

Reference - https://www.educba.com/spark-broadcast/